src_kt/java/arcs/storage/api/ArcsSingleton.kt

```diff
@@ -0,0 +1,337 @@
+import kotlinx.coroutines.Job
+import kotlinx.coroutines.async
+import kotlinx.coroutines.launch
+import kotlinx.coroutines.sync.Mutex
+import kotlinx.coroutines.sync.withLock
+import kotlin.coroutines.CoroutineContext
+import kotlin.coroutines.coroutineContext
+
+/**
+ * Creates an [ArcsSingleton] which manages a reference to a [RawEntity] object located at the given
+ * [storageKey], whose schema is defined by the supplied [schema].
+ *
+ * **Note:** By supplying your [coroutineContext], any bindings to the storage layer are released
+ * when the context's primary job is completed, thus avoiding a memory leak.
+ */
+@Suppress("FunctionName")
+@ExperimentalCoroutinesApi
+fun ArcsSingleton(
+    storageKey: ReferenceModeStorageKey,
+    schema: Schema,
+    existenceCriteria: ExistenceCriteria = ExistenceCriteria.MayExist,
+    coroutineContext: CoroutineContext = Dispatchers.Default
+): ArcsSingleton<RawEntity, RefModeStoreData.Singleton, RefModeStoreOp.Singleton> {
+    val storeOpts = StoreOptions<RefModeStoreData.Singleton, RefModeStoreOp.Singleton, RawEntity>(
+        storageKey = storageKey,
+        type = SingletonType(EntityType(schema)),
+        existenceCriteria = existenceCriteria,
+        mode = StorageMode.ReferenceMode
+    )
+    return ArcsSingleton(
+        store = Store(storeOpts),
+        toStoreData = { RefModeStoreData.Singleton(it) },
+        toStoreOp = {
+            when (it) {
+                is CrdtSingleton.Operation.Update -> RefModeStoreOp.SingletonUpdate(it)
+                is CrdtSingleton.Operation.Clear -> RefModeStoreOp.SingletonClear(it)
+                else -> throw IllegalArgumentException("Invalid operation type: $it")
+            }
+        },
+        coroutineContext = coroutineContext
+    )
+}
+
+/**
+ * Creates an [ArcsSingleton] which manages a reference to a [ReferencablePrimitive] object of
+ * type <T> located at the given [storageKey].
+ *
+ * **Note:** By supplying your [coroutineContext], any bindings to the storage layer are released
+ * when the context's primary job is completed, thus avoiding a memory leak.
+ */
+@Suppress("FunctionName")
+@ExperimentalCoroutinesApi
+inline fun <reified T> ArcsSingleton(
+    storageKey: StorageKey,
+    existenceCriteria: ExistenceCriteria = ExistenceCriteria.MayExist,
+    coroutineContext: CoroutineContext = Dispatchers.Default
+): ArcsSingleton<ReferencablePrimitive<T>,
+    CrdtSingleton.Data<ReferencablePrimitive<T>>,
+    CrdtSingleton.IOperation<ReferencablePrimitive<T>>> {
+
+    require(ReferencablePrimitive.isSupportedPrimitive(T::class)) {
+        "Unsupported type: ${T::class}"
+    }
+
+    val storeOps = StoreOptions<
+        CrdtSingleton.Data<ReferencablePrimitive<T>>,
+        CrdtSingleton.IOperation<ReferencablePrimitive<T>>,
+        ReferencablePrimitive<T>>(
+        storageKey = storageKey,
+        // TODO: this is wrong. There should probably be a PrimitiveType?
+        type = SingletonType(CountType()),
```

Comment:
Should these be indented one more level?

Reply:
This is what `ktlint` wanted... I think it looks a little weird because the generics' type arguments needed to wrap.